neo/io/nwbio.py

```diff
@@ -0,0 +1,810 @@
+    if "CurrentClamp" in timeseries_class.__name__:
+        expected_units = pq.volt
+    elif "VoltageClamp" in timeseries_class.__name__:
+        expected_units = pq.ampere
+    else:
+        # todo: warn that we don't handle this subclass yet
+        expected_units = signal.units
+    return float((signal.units/expected_units).simplified.magnitude), expected_units
+
+
+def time_in_seconds(t):
+    return float(t.rescale("second"))
+
+
+def _decompose_unit(unit):
+    """
+    Given a quantities unit object, return a base unit name and a conversion factor.
+
+    Example:
+
+    >>> _decompose_unit(pq.mV)
+    ('volt', 0.001)
+    """
+    assert isinstance(unit, pq.quantity.Quantity)
+    assert unit.magnitude == 1
+    conversion = 1.0
+
+    def _decompose(unit):
+        dim = unit.dimensionality
+        if len(dim) != 1:
+            raise NotImplementedError("Compound units not yet supported")  # e.g. volt-metre
+        uq, n = list(dim.items())[0]
+        if n != 1:
+            raise NotImplementedError("Compound units not yet supported")  # e.g. volt^2
+        uq_def = uq.definition
+        return float(uq_def.magnitude), uq_def
+    conv, unit2 = _decompose(unit)
+    while conv != 1:
+        conversion *= conv
+        unit = unit2
+        conv, unit2 = _decompose(unit)
+    return list(unit.dimensionality.keys())[0].name, conversion
+
+
+def _recompose_unit(base_unit_name, conversion):
+    """
+    Given a base unit name and a conversion factor, return a quantities unit object
+
+    Example:
+
+    >>> _recompose_unit("ampere", 1e-9)
+    UnitCurrent('nanoampere', 0.001 * uA, 'nA')
+
+    """
+    if conversion not in prefix_map:
+        raise ValueError(f"Can't handle this conversion factor: {conversion}")
+    unit_name = prefix_map[conversion] + base_unit_name
+    if unit_name[-1] == "s":  # strip trailing 's', e.g. "volts" --> "volt"
+        unit_name = unit_name[:-1]
+    return getattr(pq, unit_name)
+
+
+class NWBIO(BaseIO):
+    """
+    Class for "reading" experimental data from a .nwb file, and "writing" a .nwb file from Neo
+    """
+    supported_objects = [Block, Segment, AnalogSignal, IrregularlySampledSignal,
+                         SpikeTrain, Epoch, Event, ImageSequence]
+    readable_objects = supported_objects
+    writeable_objects = supported_objects
+
+    has_header = False
+    support_lazy = True
+
+    name = 'NeoNWB IO'
+    description = 'This IO reads/writes experimental data from/to an .nwb dataset'
+    extensions = ['nwb']
+    mode = 'one-file'
+
+    is_readable = True
+    is_writable = True
+    is_streameable = False
+
+    def __init__(self, filename, mode='r'):
+        """
+        Arguments:
+            filename : the filename
+        """
+        if not have_pynwb:
+            raise Exception("Please install the pynwb package to use NWBIO")
+        if not have_hdmf:
+            raise Exception("Please install the hdmf package to use NWBIO")
```

Comment:
hdmf has been a requirement of pynwb for a long time, so I think having pynwb installed but not having hdmf installed would be unlikely.

Reply:
I had some issues with the hdmf package a long time ago. It is for this reason that I added this additional condition. If now hdmf is a requirement of pynwb, these two lines are no longer needed.

Reply:
I agree I think we can remove this have_hdmf, no ?

Reply:
We could, but I don't think it causes any harm, so I suggest we leave it for now so we can focus on getting the release finished.